nighres/brain/dots_segmentation.py

```diff
@@ -0,0 +1,721 @@
+    Oriented Tract Segmentation (DOTS) algorithm.
+    
+    Parameters
+    ----------
+    tensor_image: niimg
+        Input image containing the diffusion tensor coefficients in the
+        following order: volumes 0-5: D11, D22, D33, D12, D13, D23
+    mask: niimg
+        Binary brain mask image which limits computation to the defined volume.
+    wm_atlas: int, optional
+        Define which white matter atlas to use. Option 1 for 23 tracts [2]_ 
+        and option 2 for 39 tracts [1]_. (default is 1)
+    max_iter: int, optional
+        Maximum number of iterations in the conditional modes algorithm.
+        (default is 20)
+    convergence_threshold: float, optional
+        Threshold for when the iterated conditonal modes algorithm is considered
+        to have converged. Defined as the fraction of labels that change during
+        one step of the algorithm. (default is 0.002)
+    g0: float, optional
+        Parameter controlling the sharpness of the membership function. (default
+        is 1/42)
+    s_I: float, optional
+        Parameter controlling how isotropic label energies propagate to their
+        neighborhood. (default is 1/42)
+    save_data: bool, optional
+        Save output data to file. (default is False)
+    overwrite: bool, optional
+        Overwrite existing results. (default is False)
+    output_dir: str, optional
+        Path to desired output directory, will be created if it doesn't exist.
+    file_name: str, optional
+        Desired base name for output files with file extension, suffixes will 
+        be added.
+    atlas_file: str, optional
+        Path to atlas directory (default is stored in ATLAS_DIR)
+
+    Returns
+    ----------
+    dict
+        Dictionary collecting outputs under the following keys
+        (suffix of output files in brackets)
+
+        * segmentation (niimg): Hard segmentation of white matter.
+        * energy (niimg): Markov Random Field energies of labels.
+        
+    Notes
+    ----------
+    Algorithm details can be found in [1]_.
+
+    References
+    ----------
+    .. [1] Bazin, Pierre-Louis, et al. "Direct segmentation of the major white 
+       matter tracts in diffusion tensor images." Neuroimage (2011)
+       doi: https://doi.org/10.1016/j.neuroimage.2011.06.020
+    .. [2] Bazin, Pierre-Louis, et al. "Efficient MRF segmentation of DTI white 
+       matter tracts using an overlapping fiber model." Proceedings of the 
+       International Workshop on Diffusion Modelling and Fiber Cup (2009)
+    """
+    
+    
+    # For external tools: dipy
+    try:
+        from dipy.core.gradients import gradient_table
+        from dipy.align.transforms import AffineTransform3D
+        from dipy.align.imaffine import MutualInformationMetric, AffineRegistration
+    except ImportError:
+        print('Error: Dipy could not be imported, it is required'
+                +' in order to run this module. \n (aborting)')
+        return None
```

Comment:
The `return None` is suboptimal, because it results in the following error when dipy isn't installed:

```
Error: Dipy could not be imported, it is required in order to run this module. 
 (aborting)
Traceback (most recent call last):
  File "examples/example_05_dots_segmentation.py", line 49, in <module>
    segmentation = dots_results['segmentation']
TypeError: 'NoneType' object is not subscriptable
```

So the important error message about dipy is a bit buried by the subsequent error message that results from returning None and then the next function trying to use None as an input. It might be solved by just doing `return` without the none. Or even break. 

(Also, this is true too for the Nipype dependency in the registration modules)

Reply:
I don't think this is a problem, because running `nighres.brain.dots_segmentation()` without DiPy prints the error message as expected and returns a `NoneType` object. The subsequent error message is due to the next line in the example where the resulting `NoneType` object is used incorrectly.